feat(DualListSelector): add support for trees and checkboxes

kmcfaul opened the pull request:
<!-- What changes are being made? Please link the issue being addressed. -->
**What**: Closes #5102 

To swap over to using trees, add `isTree` to the `DualListSelector` and update the data structure accordingly. The data is expected to have a `text` attribute to display the content of the item, and this property is used to control selection/checking/etc. The structure is similar to TreeView, but uses `text` in place of `name`. It might be better to swap to using an `id` prop instead of `text`, I haven't decided.

The prop `hasChecks` added to `DualListSelector` will add checkboxes to every option. Alternately, adding a `hasCheck` property to an option will give the individual option a checkbox. Badges are added individually at the moment with a `hasBadge` property added to an option, but if needed a similar apply-to-all property can be added.

Keyboard interaction will require some thought, and maybe requires a follow up issue, because TreeView used tabbing to navigate and DualListSelector skips over tabbing through the options.

patternfly-build commented:
PF4 preview: https://patternfly-react-pr-5207.surge.sh

mcarrano commented:
@mmenestr can you take a look at this? Not sure exactly how this should be working relative to your design. Thanks.

kmcfaul commented:
I tried to follow TreeView mostly, but had to make some assumptions about how things like the checkboxes would behave in a DualListSelector context (as selectors instead of being separate in TreeView), so any feedback would be great on interaction.

mmenestr commented:
It looks pretty good! The only thing I notice is that on nodes with no expansion, clicking on the item puts it into a selection state, but doesn't check the checkbox. It should either check the checkbox AND put the whole thing in a selected state, or just check the checkbox, and keep the text and background the same. Basically, there's a weird intermediate state that shouldn't exist! See image:
![image](https://user-images.githubusercontent.com/51165119/101068165-61e0f080-3566-11eb-8dce-e99a5c7562e7.png)

The other thing i'm iffy about is the fact that clicking on a row **with** an expansion both selects it AND opens it. I don't think that should be the behavior and I wonder if we could do one of two things:

**Option 1:** Separate the row into two "click areas", and make it so that if they click anywhere left of the checkbox - so on the arrow or left of it, the row expands, but if they click on the checkbox or anything right of the checkbox, then it selects the item without expanding? 

**Option 2:** Decide that users can't select something unless they click on the checkbox itself, and any other click on the row would either expand the row (for rows with carets) or do nothing (for rows without). This might be the simplest and more intuitive option.

Curious to hear your thoughts on what you'd expect^ @mcarrano @kmcfaul

kmcfaul commented:
I think that if the user wants the checkboxes, it would make sense for them to replace the row selection. At least I'm not sure if there's a use case for checkboxes outside of selection.

I can implement either way though. I already separated expansion from selection for the toggle and can do the same for the whole row (so expansion is controlled only by the toggle button, not the row click).

kmcfaul commented:
Realized I didn't update the search functionality with trees, so working on including that as well as a basic integration demo/test.

nicolethoen commented:
This is cool! I'd say that if they are using checkboxes, then possibly the checkboxes should just toggle alongside the selected state? so the checkbox just emphasizes the selected state?
and a partially checked checkbox would not imply the selected state?

mmenestr commented:
@nicolethoen I think that makes sense! 
And as for the expansion vs selection thing, I just got a second designer opinion and we're both leaning towards keeping it simple and going with option 2 which would be:
-  if the user clicks on the checkbox, the row gets selected
- if the user clicks anywhere else (arrow or general row outside of checkbox) the folder expands. 
- But there should never be a place where a click causes both a selection AND a folder expansion

mcarrano commented:
Agree with @mmenestr above. I think that this also means that clicking outside of the checkbox on a row without children has no effect (the user must use the checkbox to select/deselect items).

kmcfaul commented:
Updated! Whenever an option has a checkbox, it will prevent the row selection and only allow the checkbox toggle. So if a user is using the tree, but not checkboxes, row selection will behave as before.

The checkbox handler in DualListSelector handles the checkbox as well as selection, and also now triggers the optional selection callback.

mcarrano commented:
The interaction feels right now @kmcfaul . Do you know if this is consistent with how we are handling the Tree view? It should behave the same, I think. I'm also wondering if it's possible to clean up the alignment issue I'm seeing below:

![Screen Shot 2020-12-03 at 4 04 24 PM](https://user-images.githubusercontent.com/6619860/101088626-f9077180-3581-11eb-950e-fccd5fa9af20.png)

Does that require a core fix?

Also, I'm not clear what the purpose of the badge is. Can you remind me of what that is intended to indicate @mmenestr ? It doesn't seem to change with anything.

kmcfaul commented:
The alignment of folders to options would probably be a core issue. @mcoker 

The badge is showing the number of direct children. This is the same as Tree View and I think what the core example showed, but I'm unsure.

It's not the same interaction, TreeView's selection and checking behaviors were separate. Selection only highlighted the option and its direct parent, while checking could be done to any option and didn't trigger selection at all. I think the context of Dual List Selector changes the behavior opposed to Tree View (which may be used for a menu or the like). But I could be wrong, should something change?

tlabaj commented:
@mcarrano @mmenestr wouldn't we want this component's interactions to be consistent with the tree view component.  See PR #https://github.com/patternfly/patternfly-react/pull/5128 for the updates to the checkbox interactions in the tree view component. 
The only way to expand the node is the click the toggle.  clicking the label or checkbox selects the item.

mcarrano commented:
@tlabaj I had the same thought, initially. But @kmcfaul makes a good point that the usage context is a little bit different. What do you think @mmenestr ? I can live with the inconsistency if we think that both components make sense in their separate contexts. Users won't think hard about this if it just works.

mmenestr commented:
Looks a lot better! Just have 2 things:
1. I'm still able to click on an item row, and it makes it grey - not sure if that's the focus state or what but it seems weird to me. 
2. The count number based on selections is wrong. I should only count items and not folders! So in this example, it should say 2 out of 6 items, not 4 out of 9 items.

@mcarrano the badge is meant to show the number of items inside the folder.

kmcfaul commented:
I think the grey background is from the hover state? And/or the focus state. It's definitely not 'selected'. @mcoker Any insight where the color is coming from? I think in TreeView the highlighting was to show placement with keyboard navigation but I'm not sure where the styling is coming from.

I'll work on updating the count number. I think I initially had it counting just the options (not folders) but it got a bit off in count and I changed it to all items. I'll see if I can figure it out.

@tlabaj @mcarrano I can update the behavior as such (toggle only expands, everything else checks the checkbox if it exists) we if want. Let me know!

mcarrano commented:
I'll defer to @mmenestr on the checkbox and toggle interactions. I think users will figure it out either way.

mmenestr commented:
Just looked at the tree view implantation and from what I can see, the only difference is that you can only expand by clicking the toggle which I'd be fine to copy for this component too.
When I click on the label in the tree view though, it doesn't get selected like you're saying @tlabaj so i'm confused! It only gets selected when I click on the checkbox.

kmcfaul commented:
Updated. 
Only the toggle will expand/collapse a folder now.

mcarrano commented:
@kmcfaul I think that the interaction makes sense now. My only issue it with the styling of selected items. Why do they have a gray background like this?

<img width="396" alt="Screen Shot 2020-12-04 at 9 17 38 AM" src="https://user-images.githubusercontent.com/6619860/101174479-ee94b880-3611-11eb-9c33-0e7021da7c12.png">

I would only expect that the background shading is used to indicate hover or focus and selection in just indicated with the blue text. I don't have the core example to compare against. Can you check this @mcoker ?

mmenestr commented:
@mcarrano that is the selected state for non-checkbox select elements, when you have a dual list with a single level, so it was just used again here. Alternatively, we could get rid of the added visual element, and only show a selection via the checkbox!

mcarrano commented:
@mmenestr To me it feels redundant to have both.

mmenestr commented:
@mcarrano ok, I don't have a strong preference either way, so I'm fine with removing all extra styling OR with just doing the blue text, with no grey background! I'll let you make the call :) The only reason for keeping it would be added emphasis - but I agree it may be redundant.

mcarrano commented:
I think just having the checkbox selected and blue text is enough. The gray fill should be reserved to indicate hover/focus IMO.

mcarrano reviewed (approved):
Looks good pending core updates. Thanks @kmcfaul .

kmcfaul commented:
Bumped core. There is a small issue that @mcoker is looking at (where the alignment of checkbox items with no children on the top level doesn't match other folders).

jschuler commented:
@kmcfaul can you describe why you can't use the existing `<TreeView>` component? Seems a bit heavy to recreate it for dual list

mcoker commented:
@jschuler I can say from the core side we talked to design and while they share a similar layout, they aren't the same component they may have different features, layouts, and don't necessarily need to change together when either is updated.

That said, I think there is enough design and functionality that maybe we have a shared component that both tree view and the dual list selector can use that give the base element and layout support, and those components modify the layout, state, etc changes unique to that component. wdyt?

jschuler commented:
@mcoker can we not build in more features/variants/modifiers into the existing TreeView component? seems like it could cause more headaches in the future if they diverge already.

kmcfaul commented:
Mostly what @mcoker said. The structure of the Tree component is similar but at least the Item is slightly different. The interaction patterns are also somewhat unique because DualListSelector is a specific use case of TreeView while TreeView is more generic. And the keyboard interaction (which needs to be defined for DualListSelector still) is entirely different since TreeView uses tabbing and DualListSelector ignores tabbing through the list entirely atm.

jschuler commented on packages/react-core/src/components/DualListSelector/DualListSelector.tsx:
```diff
@@ -217,13 +322,131 @@ export class DualListSelector extends React.Component<DualListSelectorProps, Dua
+          .filter(item => filterTreeItems(item as DualListSelectorTreeItemData, [text]));
+        const flatSelectedItems = flattenTreeWithFolders(selectedOptionTree as DualListSelectorTreeItemData[]);
 
+        if (selectedOptions.includes(text)) {
```
seems this might be a bit brittle? considering you could have multiple branches with the same named nodes? clicking one would auto-check the other
<img width="375" alt="Screen Shot 2020-12-08 at 10 06 39 AM" src="https://user-images.githubusercontent.com/869106/101501055-4185ad00-393d-11eb-8a90-6b9f9eaecf6a.png">

jschuler replied:
Another example: In this case, selecting `Option Z` highlights both parent folders. Maybe this is an edge case but perhaps we shouldn't key off of text for identifying a node
<img width="369" alt="Screen Shot 2020-12-08 at 10 12 36 AM" src="https://user-images.githubusercontent.com/869106/101501875-439c3b80-393e-11eb-9fb8-1c2bc7c02b1a.png">

kmcfaul replied:
Good point. Maybe I should key off of `id` or a similar prop instead.

jschuler commented on packages/react-core/src/components/DualListSelector/DualListSelector.tsx:
```diff
@@ -60,13 +70,21 @@ export interface DualListSelectorProps {
   /** Id of the dual list selector. */
   id?: string;
+  /* Flag indicating if the dual list selector uses trees instead of simple lists */
+  isTree?: boolean;
```
this might be helpful
https://quicktype.io/typescript/

kmcfaul replied:
availableOptions and chosenOptions both have `React.ReactNode[]` typing, I'm not sure what you mean here. Do you mean for me to change it to `React.ReactNode[] | DualListSelectorTreeItemData[]`?

jschuler replied:
yes, so TS consumers can get hints and auto-completion as to what is supported in their options definition array when creating a tree configuration

kmcfaul replied:
I'll also add this in the description for clarity.

jschuler commented on packages/react-core/src/components/DualListSelector/DualListSelector.tsx:
```diff
@@ -299,13 +522,40 @@ export class DualListSelector extends React.Component<DualListSelectorProps, Dua
     const availableOptionsStatusToDisplay =
-      availableOptionsStatus || `${availableOptionsSelected.length} of ${availableOptions.length} items selected`;
+      availableOptionsStatus ||
+      (isTree
```
this is not i18n friendly (feel free to open followup issues as needed)

kmcfaul replied:
The custom string `availableOptionStatus` should overwrite either of the default strings. What about it isn't i18n friendly?

jschuler replied:
oh sorry, i didn't see that

tlabaj reviewed (changes requested):
Is the dual list selector tree only intended to be used with checks?  If not, it might be worth adding an example without checks and verifying that interactions work as expected.
cc @mcarrano

mmenestr commented:
@tlabaj the dual list with **multiple levels** is only intended to be used with checkboxes yes.
But, a dual list with only a single level of options does not need checks though (ie: it's just a straight list of options with no nesting).

kmcfaul commented on packages/react-core/src/components/DualListSelector/DualListSelector.tsx:
```diff
@@ -86,14 +104,19 @@ export class DualListSelector extends React.Component<DualListSelectorProps, Dua
     removeSelectedAriaLabel: 'Remove selected',
     removeAllAriaLabel: 'Remove all'
   };
+  private originalCopy = this.props.availableOptions;
```
I think I do need it because availableOptions can be re-passed in after a list update (like add/remove). To maintain the overall order I need an initial array.

kmcfaul replied:
Would it be better to make a separate prop to use instead? Like `initialOptions`?

jschuler replied:
oh, I am thinking that originalCopy is just a pointer to prop.availableOptions, so any changes to one should reflect in the other. Just would be cleaner not to have 3 things in code tracking availableOptions is all. But it's only a nit so if it causes issues it's okay the way it is
```
originalCopy
props.availableOptions
state.availableOptions
```

kmcfaul commented on packages/react-core/src/components/DualListSelector/DualListSelector.tsx:
```diff
@@ -173,6 +206,39 @@ export class DualListSelector extends React.Component<DualListSelectorProps, Dua
+    this.setState(prevState => {
+      // Remove selected available nodes from current available nodes
+      const newAvailable = prevState.availableOptions
+        .map(opt => Object.assign({}, opt))
```
It's making a copy to prevent array mutation.

jschuler replied:
array filter is already non-mutating and creates a new array. I'd go through and remove unnecessary usages of the map

jschuler replied:
btw, I often reference https://lorenstewart.me/2017/01/22/javascript-array-methods-mutating-vs-non-mutating/

kmcfaul replied:
`filter` doesn't mutate, but I think it's because of the recursion that it's needed. I tried removing some of the `.map` but then I start losing options when it filters/checks/selects.

redallen replied:
@kmcfaul Can you use `Array.from(prevState.availableOptions)` instead? It's a little more explicit about the copy.

kmcfaul commented on packages/react-core/src/components/DualListSelector/DualListSelector.tsx:
```diff
@@ -201,7 +267,46 @@ export class DualListSelector extends React.Component<DualListSelectorProps, Dua
   };
 
-  onOptionSelect = (e: React.MouseEvent | React.ChangeEvent, index: number, isChosen: boolean) => {
+  removeTreeSelected = () => {
```
They are similar in code but the arrays they target are different. I would personally rather keep them separate than have if/else logic inside to change around the targeted arrays, it seems more clearer to me this way. But am opening to change it if that's the consensus.

kmcfaul commented on packages/react-core/src/components/DualListSelector/DualListSelector.tsx:
```diff
@@ -217,13 +322,131 @@ export class DualListSelector extends React.Component<DualListSelectorProps, Dua
       };
     });
+    const onOptionSelect = this.props.onOptionSelect;
+    if (onOptionSelect) {
```
Ah yeah I see a couple instances where this could be shortened to `props.method && props.method(e)`. Will update.

kmcfaul commented on packages/react-core/src/components/DualListSelector/DualListSelectorTree.tsx:
```diff
@@ -0,0 +1,149 @@
+  /** Text of the option */
+  text: string;
+  /** Checked state of the option */
+  checked: boolean;
```
Yeah I can rename it.

kmcfaul commented on packages/react-core/src/components/DualListSelector/DualListSelectorTree.tsx:
```diff
@@ -0,0 +1,149 @@
+  /** Internal callback to pass this ref up to the parent. */
+  sendRef?: (optionRef: React.ReactNode, index: number) => void;
+  /** Internal field used to keep track of the order of filtered options. */
+  filteredIndex?: number;
```
I'm removing these props, since the Tree version isn't building a ref array (as of yet). They were holdovers from the initial DualListSelectorItem I worked off of.

Also, only `DualListSelector` is exported so I'm not sure if it matters, since a user can't access the sub-components?

redallen replied:
@tlabaj Not sure if this applies anymore, but the way to hide props is to start the comment with `@hide` like `/** @hide Internal field consumers need not worry about */`.

kmcfaul commented:
Updated with PR feedback.

- Tree now uses `id` for filtering/lookup etc (except in search function which still looks at text), to prevent duplicate selections when options share the same text.
- `checked` renamed to `isChecked`.
- Removed some unused props from the Tree/TreeItem. These were holdovers from the base DualListSelector code I started from, but are not used by the tree right now.
- Condensed some optional callback checking & triggering to inline.
- Re-enabled eslint no-unused props check.

tlabaj commented:
> @tlabaj the dual list with **multiple levels** is only intended to be used with checkboxes yes.
> But a dual list with only a single level of options does not need checks (ie: when it's just a straight list of options with no nesting).

@kmcfaul  if this is the case, is there really a need to have the `hasCheck` prop and all the logic associated with that?

jessiehuff reviewed (commented):
For keyboard interaction, I think it would make the most sense to follow the [authoring practices for the treeview component](https://www.w3.org/TR/wai-aria-practices-1.2/#TreeView). I believe we should be following the pattern of a treeview that supports multi-selection, using left/right arrow keys to handle expansion, down/up arrow keys to handle focus between options, and Enter/Space for node selection. See the [keyboard interaction](https://www.w3.org/TR/wai-aria-practices-1.2/#keyboard-interaction-23) for all specifics. Also, I'm currently unable to use the keyboard to expand or select options.
Interestingly enough, I was able to perform some interactions with VO (checking the checkbox), but I was not able to expand or select options. I'm thinking that the interaction pattern between keyboard and screen reader might be linked here. I imagine this may be similar to [this open issue](https://github.com/patternfly/patternfly-react/issues/5149).

kmcfaul commented:
@tlabaj If that's the case then I can remove that prop and always include checkboxes

jschuler commented:
Can you also remove `hasChecks` from the DualListSelectorPane? And a rebase :)

jessiehuff commented:
Added an [issue](https://github.com/patternfly/patternfly-react/issues/5231) for the keyboard/screen reader expectations. 🙂

tlabaj commented on packages/react-core/src/components/DualListSelector/DualListSelectorTreeItem.tsx:
```diff
@@ -0,0 +1,154 @@
+  /** Flag indicating this option is expanded by default. */
+  defaultExpanded?: boolean;
+  /** Flag indicating this option has a checkbox */
+  hasCheck?: boolean;
```
Shouldn't all the tree items have checks?  If so, this should not be needed.

kmcfaul replied:
I thought to leave it in just in case a user wanted to make something un-checkable (like a folder), but I can remove it.

tlabaj replied:
@mmenestr @mcarrano thoughts on this?  Do we want to allow them to not have the checkbox?

mcarrano replied:
I don't think so. It's not something I've seen. What do you think @mmenestr?

kmcfaul replied:
I'll remove it for now then.

tlabaj reviewed (approved):
Looks great!

mmenestr commented:
Looks good! And like Matt said, every item should have a checkbox in this multi level dual list. The only thing i'd say now is that maybe we should revert the badge color back to the default grey for badges, since we no longer have the background hover state that would clash with it.

nicolethoen commented:
There is a lot of conversation here, so this may be a done decision.
I'd just love some clarity.

In the tree view example, my mouse becomes a pointer and the item gets focus when I click on the text of an item, but nothing else happens. so it's focused, but not selected or checked? Why is that? Should clicking when the mouse looks like a pointer have a more obvious side effect?
![Dec-09-2020 15-13-49](https://user-images.githubusercontent.com/26305082/101682259-60656b80-3a31-11eb-815f-9ce672e4ab0b.gif)


And i'm not sure if the enter key is doing anything? or the arrow keys? 
In the previous examples, you can tab to the list, then need to use the arrow keys to navigate the list items. So the tree list implementation is definitely different than the other examples. Is that on purpose?

Sorry if this is repeat.

kmcfaul commented:
I'm not sure about the pointer. Perhaps it's a holdover when row selection did something and now I need to remove a class maybe? @mcoker Any thoughts?

Keyboard navigation has a follow up issue #5231, it needs to be defined and added.

nicolethoen commented:
we are sure we don't want clicking the row to check the box?

mcoker commented:
@nicolethoen @kmcfaul agree, the pointer cursor and background color change on hover should be paired with some kind of interactivity. It seems like we should be consistent between tree view and the dual list selector?

It seems like clicking anywhere in an item should perform the primary action for that item:

* Clicking a regular item selects it
* Clicking a regular expandable item expands it
* Clicking a checkbox item - should check it
* Clicking an expandable checkbox item - not sure? What's the primary action here?

wdyt @mcarrano?

We also made some structural updates to the tree view that I think we could make in this component, too, including making the text next to the checkbox a `<label>` (with a `for` attr that matches the `id` of the `input[type="check"]`), which will enable the checkbox to toggle the checked state when clicking on the label.

mcarrano commented:
I agree with @mcoker above. Clicking anywhere on the item should select or de-select it by toggling the checkbox state. The only way to expand or collapse an item is to click on the expansion caret. I'm pretty sure that's consistent with what we said before. Can we make that change?

kmcfaul commented:
Added the row click back in, toggling the checkbox (and therefore selection) just as the checkbox toggle would.

mcarrano reviewed (approved):
This is great. Thanks for making that change @kmcfaul !

mcoker reviewed (approved):
Looks super! 🤩

mmenestr commented:
before we merge @mcarrano, what do you think about changing the badge color to the lighter default grey? As opposed to the darker grey we initially switched so that it would not clash with the hover background (that we no longer have)

mcarrano commented:
@mmenestr I think that we still do have the light gray background on hover (but not select). In any case, I'd like to get this merged this morning, if possible, so let's freeze the design for now. If there are follow on updates needed, they can be made in a future release since the component is still considered beta. Thanks.

tlabaj reviewed (approved):
LGTM

jschuler reviewed (approved):
LGTM
